### PR TITLE
Feature: EGDS support to enable EDS patch

### PR DESCRIPTION
**Pull request by seflerZ**
This pull request introduces the EGDS support which is discussed [here](https://github.com/istio/istio/issues/19287).

EGDS boosts network transport efficiency when a small piece of endpoints changed within EDS. The design doc is [here](https://docs.google.com/document/d/1I5B385RMV4m5DohowKdRIdY5s6Xn97_1VdyjjqyLwlQ/edit). This feature also involves implementation on Enovy side. You can find relative design docs and issues in the [issue](https://github.com/istio/istio/issues/19287). The Envoy side PR is [here](https://github.com/envoyproxy/envoy/pull/10023). Due to small piece of changes will be applied, Envoy will save 60%+ CPU usage when endpoints are likely to change constantly according to our test.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -10,6 +10,10 @@ replace github.com/spf13/viper => github.com/istio/viper v1.3.3-0.20190515210538
 
 replace github.com/docker/docker => github.com/docker/engine v1.4.2-0.20191011211953-adfac697dc5b
 
+replace github.com/envoyproxy/go-control-plane => gitlab.alibaba-inc.com/AliMesh/go-control-plane v0.9.2-0.20200218115735-e6560592fa4e
+
+replace istio.io/api => gitlab.alibaba-inc.com/one-mesh/api v0.0.0-20200215150939-aa28515cd791
+
```
Comment (seflerZ) [inline go.mod]: Don't care about any dependency starts with "alibaba-inc.com". I'll change it to opensource version before merging and change them to open source version once the changes are applied in opensource libs.

**Comment by rshriram** [comment]
cc @ramaraochavali / @hzxuzhonghu / @howardjohn this is your territory..

File: pilot/pkg/proxy/envoy/v2/eds.go [hunk 2]
```diff
@@ -634,117 +657,194 @@ func (s *DiscoveryServer) loadAssignmentsForClusterIsolated(proxy *model.Proxy,
...
-func (s *DiscoveryServer) generateEndpoints(
-	clusterName string, proxy *model.Proxy, push *model.PushContext, edsUpdatedServices map[string]struct{},
-) *xdsapi.ClusterLoadAssignment {
-	_, _, hostname, _ := model.ParseSubsetKey(clusterName)
-	if edsUpdatedServices != nil {
-		if _, ok := edsUpdatedServices[string(hostname)]; !ok {
-			// Cluster was not updated, skip recomputing. This happens when we get an incremental update for a
-			// specific Hostname. On connect or for full push edsUpdatedServices will be empty.
-			return nil
+func (s *DiscoveryServer) loadAssignmentFailback(clusterName string, groupName string,
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/eds.go]: Fallback

File: pilot/pkg/model/push_context.go [hunk 3]
```diff
@@ -284,15 +286,27 @@ func (first *PushRequest) Merge(other *PushRequest) *PushRequest {
...
+
+		if first.EgdsUpdates != nil || other.EgdsUpdates != nil {
+			merged.EgdsUpdates = make(map[string]struct{})
+			for update := range first.EgdsUpdates {
+				merged.EgdsUpdates[update] = struct{}{}
+			}
+			for update := range other.EgdsUpdates {
+				merged.EgdsUpdates[update] = struct{}{}
+			}
+		}
```
Comment (rshriram) [inline pilot/pkg/model/push_context.go]: Why do you need to have another copy same state in the form of EdgsUpdates? Why can't you use the existing EdsUpdates ?

**Comment by ramaraochavali** [comment]
@seflerZ  I see that EGDS API in Envoy is still being reviewed https://github.com/envoyproxy/envoy/pull/10023 - Can we please get the API finalized/approved there so that we can review the implementation here?

**Comment by seflerZ** [comment]
Thank you so much for your attention. The Envoy API on EGDS was designed and implemented weeks ago. There will not be much difference between current and release version. I posted the PR on Pilot because I think it’s better to have the basic EGDS processes get reviewed in advance. The PR on Envoy side may take some time based on the community review work, but don’t worry I’ll push the work through until merged.

发送自 Windows 10 版邮件应用

发件人: Rama Chavali
发送时间: 2020年2月21日 16:49
收件人: istio/istio
抄送: Sefler Zhou; Mention
主题: Re: [istio/istio] Feature: EGDS support to enable EDS patch (#21254)

@seflerZ I see that EGDS API in Envoy is still being reviewed envoyproxy/envoy#10023 - Can we please get the API finalized/approved there so that we can review the implementation here?

File: pilot/pkg/model/push_context.go (hunk 3 above)
Comment (seflerZ) [reply pilot/pkg/model/push_context.go]: The 'EgdsUpdates' stores the changed groups while the 'EdsUpdates' only has the changed service. For example, if we a service named A has 100 endpoints and we slice it into 5 groups. With 15 endpoints change will probably involve 2 or more groups.  So the data in 'EdsUpdates' will be '{A: {}}' while 'EgdsUpdates' will be {'A-G1':{},'A-G2':{},'A-G5':{}}.

> Why do you need to have another copy same state in the form of EdgsUpdates? Why can't you use the existing EdsUpdates ?

The 'EgdsUpdates' stores the changed groups while the 'EdsUpdates' only has the changed service. For example, if we a service named A has 100 endpoints and we slice it into 5 groups. With 15 endpoints change will probably involve 2 or more groups.  So the data in 'EdsUpdates' will be '{A: {}}' while 'EgdsUpdates' will be {'A-G1':{},'A-G2':{},'A-G5':{}}.

File: pilot/pkg/proxy/envoy/v2/eds.go (hunk 2 above)
Comment (seflerZ) [reply pilot/pkg/proxy/envoy/v2/eds.go]: okay, thanks.

**Comment by seflerZ** [comment]
Close due to Envoy has no support

**Pull request closed without merging**